### ISSUE TITLE
[Bug]: [help wanted] MoE + TP + custom allreduce bug

### ISSUE THREAD

**Issue by youkaichao** [vllm-project/vllm#9774]
### Your current environment

main branch, H100

### 🐛 Describe the bug

when I try this simple command `vllm serve allenai/OLMoE-1B-7B-0924 -tp 2` on the main branch, it hits an error:

`Failed: Cuda error /workspace/csrc/custom_all_reduce.cuh:336 'invalid argument'`

removing `-tp` works, and disabling custom allreduce also works.

not sure what's happening. if anyone is familiar with the moe code, please help.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by youkaichao** [comment]
@cedonley thanks for the great information! can you try to confirm which commit leads to this bug? we have all the wheels from the main branch that's ready to directly install, see https://docs.vllm.ai/en/latest/getting_started/installation.html#install-the-latest-code .

**Comment by cedonley** [comment]
@youkaichao Sure. I was just able to confirm the following results:

commit# 3cb07a36a20f9af11346650559470d685e9dc711 (after #9588) - FAILS
commit# 8549c82660cfa59a13cccd622f8afcc29cbd4281 (just before) - SUCCESS

The invalid argument error is coming from a call to cudaIpcOpenMemHandle. Did some various internet searching to see what might have changed here between CUDA 12.1 and 12.4 (which is the other factor introduced in #9588) and coming up empty at the moment.

**Comment by youkaichao** [comment]
might be caused by the new expandable segment:

https://github.com/pytorch/pytorch/blob/18ad44e830e6f885ee81c9ccf65f67c2fa05b9eb/c10/cuda/CUDACachingAllocator.cpp#L348

**Comment by youkaichao** [comment]
it looks `expandable_segments` is `False` by default.

we need to figure out the memory allocation strategy in pytorch 2.5 .

if you have time, you can also compile yourself, and call cuda APIs to get the pointer's attributes, that should tell you more information.

**Comment by cedonley** [comment]
Did a little more analysis last night and it's clear that the flow *should* be working. Here's what I see:

**custom_all_reduce.py**
- allocates the IPC metadata using PyTorch storage (and I see it come back)
- handles/offsets look fine, rank/world size is correct
- it then passes this all as parameters to init_custom_ar
- *the error occurs before it registers the buffers and returns*

**custom_all_reduce.cu**
- _init_custom_ar_  - sees the correct parameters, makes a copy of the handles allocated by PyTorch in the python code, and then passes this into the _CustomAllreduce_ function defined in custom_all_reduce.cuh

**custom_all_reduce.cuh**
- _CustomAllreduce_ - Immediately tries to open the handles, which results in the error we see

Not finding any changes in recent PyTorch releases related to how storage is allocated. The metadata that is passed is allocated on python with essentially (simplified):
```
meta = torch.zeros(ops.meta_size() + max_size,
                                dtype=torch.uint8,
                                device=self.device)
data = meta.untyped_storage()._share_cuda_()
shard_data = (
            data[1],  # ipc handle to base ptr
            data[3],  # offset of base ptr
        )
handles, offsets = self._gather_ipc_meta(shard_data)
```

I see both the handles and offsets come back and don't see any issues with them.

I tried changing the way processes were spawned, etc... to see if perhaps there was a change to multiprocessing that created a side effect, etc... I'm stumped... Someone with stronger CUDA debugging skills is going to need to have a look at this.

**Comment by youkaichao** [comment]
thanks for the investigation! let me ask some cuda experts.

**Comment by youkaichao** [comment]
I think https://github.com/pytorch/pytorch/pull/130888 might be the culprit .

**Comment by youkaichao** [comment]
okay, they changed the binary format of the ipc handle in https://github.com/pytorch/pytorch/pull/130890
